bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -788,89 +852,340 @@ private long readLastLogId(File f) {
+        }
+
+        @Override
+        public void checkpoint() throws IOException {
+            flushRotatedLogs();
+        }
+
+        @Override
+        public synchronized void rollLogs() throws IOException {
+            createNewLog(UNASSIGNED_LEDGERID);
+        }
+
+        public long getCurrentLogId() {
+            BufferedLogChannel currentActiveLogChannel = activeLogChannel;
+            if (currentActiveLogChannel != null) {
+                return currentActiveLogChannel.getLogId();
+            } else {
+                return EntryLogger.UNINITIALIZED_LOG_ID;
+            }
+        }
+
+        @Override
+        public void flushCurrentLogs() throws IOException {
+            BufferedLogChannel currentActiveLogChannel = activeLogChannel;
+            if (currentActiveLogChannel != null) {
+                /**
+                 * flushCurrentLogs method is called during checkpoint, so
+                 * metadata of the file also should be force written.
+                 */
+                flushCurrentLog(currentActiveLogChannel, true);
+            }
+        }
+
+        @Override
+        public void closeCurrentLogs() throws IOException {
+            if (activeLogChannel != null) {
+                closeFileChannel(activeLogChannel);
+            }
+        }
+
+        @Override
+        public void forceCloseCurrentLogs() {
+            if (activeLogChannel != null) {
+                forceCloseFileChannel(activeLogChannel);
+            }
+        }
+
+        @Override
+        public void prepareEntryMemTableFlush() {
+            logIdBeforeFlush = getCurrentLogId();
+        }
+
+        @Override
+        public boolean commitEntryMemTableFlush() throws IOException {
+            long logIdAfterFlush = getCurrentLogId();
+            /*
+             * in any case that an entry log reaches the limit, we roll the log
+             * and start checkpointing. if a memory table is flushed spanning
+             * over two entry log files, we also roll log. this is for
+             * performance consideration: since we don't wanna checkpoint a new
+             * log file that ledger storage is writing to.
+             */
+            if (reachEntryLogLimit(activeLogChannel, 0L) || logIdAfterFlush != logIdBeforeFlush) {
+                LOG.info("Rolling entry logger since it reached size limitation");
+                rollLogs();
+                return true;
+            }
+            return false;
         }
-        // move the leastUnflushedLogId ptr
-        leastUnflushedLogId = flushedLogId + 1;
     }
 
+    /**
+     * Flushes all rotated log channels. After log channels are flushed,
+     * move leastUnflushedLogId ptr to current logId.
+     */
+    void checkpoint() throws IOException {
+        entryLogManager.checkpoint();
+    }
+
+
+
     public void flush() throws IOException {
-        flushRotatedLogs();
-        flushCurrentLog();
+        entryLogManager.flushCurrentLogs();
```

Comment:
for single-log manager, it is better to flush `rotated logs` before flushing `current logs`, because the way how it rotates files and do the checkpoint. I would prefer keeping existing logic unchanged to reduce the risks.

besides that, why not provide `flush` method in EntryLogManager? so different entry log manager implementation can implement its own flushing logic rather than exposing this knowledge to EntryLogger. If you provide `flush` method, it is possible that you don't need to expose `flushCurrentLogs` and `flushRotatedLogs`.

Reply:
moved flushCurrentLogs and flushRotatedLogs from EntryLogManager interface to EntryLogManagerBase and added flush method in interface.

Reply:
btw, regarding order of flushRotatedLogs and flushCurrentLogs, I responded earlier as well - 

i think it is incorrect to call flushrotatedlogs first and then flushcurrentlogs. Consider this scenario - if in between flushRotatedLogs, flushCurrentLogs calls current activelog is rotated and added to rotatedlogs list then we would miss flushing that log in this current flush call. So it has to be first flushCurrentLogs and then flushRotatedLogs.

Reply:
> Consider this scenario - if in between flushRotatedLogs, flushCurrentLogs calls current activelog is rotated and added to rotatedlogs list then we would miss flushing that log in this current flush call. So it has to be first flushCurrentLogs and then flushRotatedLogs.

Can you add a test case for that? I don't see how this can happen in single-entry-log implementation.